### DISCUSSION TITLE
Copy and deep copy of array

### DISCUSSION THREAD

**Discussion by zcbenz** [ml-explore/mlx#1029] [Q&A] [answered]
My understanding of below code is, it uses `copy` op to create a new array with a new ArrayDesc, with the tracing turned on.
https://github.com/ml-explore/mlx/blob/d0dbfe0b97b95bc29009e7534f7ff734fb679801/mlx/transforms.cpp#L186-L193

What does the "Does not do a deep copy" comment mean?

Is there a concept of deep copy in MLX? The `__deepcopy__` operator in python is simply implemented as `array(self)`.

**Comment by awni** [comment, marked as answer]
The comment there just indicates that `copy(p, s)` is not going to make a new data buffer to back it's output, it will just be a view of the same data as `p`.

There is no real "deep copy" in MLX in the sense of an operation which forces a new data buffer to be allocated and copied into. We have no plans to expose one. Data buffers are managed by primitives and the runtime which is at a lower level and having complete control over that is useful.
